### DISCUSSION TITLE
Unclear how to add source from memory

### DISCUSSION THREAD

**Discussion by michielbdejong** [comunica/comunica#451] [Q&A] [answered]
#### Issue type:

- :question: Question

____
#### Question:

From https://comunica.readthedocs.io/en/latest/actors/algebra/#context-entries it seems that possible values for sources in the second parameter of [`myEngine.query`](https://github.com/comunica/comunica/tree/master/packages/actor-init-sparql#usage-within-application) are restricted to remote sources and files. But I want to parse a local Turtle file from memory, after retrieving it from the persistence layer (which is not necessarily the filesystem, could be a Redis backend, etc.). Is there a way to use a Turtle document or a triple stream as a query source?

**Comment by rubensworks** [comment, marked as answer]
The default Comunica config does not allow querying from local files for security reasons.

However, as querying from local files is a typical use case, we do offer a Comunica config that allows querying from local files: [Comunica SPARQL file](https://github.com/comunica/comunica/tree/master/packages/actor-init-sparql-file#readme). Does this work for your use case?

**Comment by michielbdejong** [comment]
I could fork that and see where it calls `fs.readFile` and then insert my string there.

**Comment by michielbdejong** [comment]
but why can't you just pass it a more generic ReadableStream instead of an `fs` file, for instance?

**Comment by rubensworks** [comment]
> but why can't you just pass it a more generic ReadableStream instead of an fs file, for instance?

Architecturally, that definitely possible. There's just no Comunica actor yet that does this.

It should be quite straightforward to implement an actor that provides RDF from a `ReadableStream`, similar to how the current [file actor](https://github.com/comunica/comunica/blob/master/packages/actor-rdf-dereference-file/lib/ActorRdfDereferenceFile.ts#L49) works now.

**Comment by RubenVerborgh** [comment]
Can't what @michielbdejong wants be realized by a) parsing the triple stream into a store b) having Comunica query over the store? I know that Comunica accepts RDF/JS in-memory stores.

For instance, just pass the readable stream through `rdflib` or `N3.Parser` and we're all good, no?

**Comment by rubensworks** [comment]
Yep, that also makes sense.
Delegating parsing to an external library or letting Comunica do the parsing by only passing a raw stream would both work. But depending on the use case @michielbdejong has, one may be better than the other.

**Comment by michielbdejong** [comment]
I need something like this:
```js
const turtleText = `@base <http://example.org/> .
@prefix foaf: <http://xmlns.com/foaf/0.1/> .
@prefix rel: <http://www.perceive.net/schemas/relationship/> .

<#green-goblin>
    rel:enemyOf <#spiderman> ;
    a foaf:Person ;    # in the context of the Marvel universe
    foaf:name "Green Goblin" .
`

const sparqlQuery = `PREFIX foaf:  <http://xmlns.com/foaf/0.1/>
SELECT ?name
WHERE {
  ?person foaf:name ?name .
}`

const result = executeQuery(turtleText, sparqlQuery)
expect(result).toEqual(`@base <http://example.org/> .
@prefix foaf: <http://xmlns.com/foaf/0.1/> .

<#green-goblin>
    foaf:name "Green Goblin" .
`)
```
Can you maybe give an example of how to do that using `N3.parser` + `Comunica`? That would really help us!

**Comment by michielbdejong** [comment]
> I know that Comunica accepts RDF/JS in-memory stores.

@rubensworks @RubenVerborgh how? https://github.com/comunica/comunica/tree/master/packages/actor-init-sparql#usage-within-application does not document that, apparently?

**Comment by rubensworks** [comment]
`@comunica/actor-init-sparql` indeed does not support this by default. You can use [`@comunica/actor-init-sparql-rdfjs`](https://github.com/comunica/comunica/tree/master/packages/actor-init-sparql-rdfjs#readme) for this. (see usage example in README)

**Comment by michielbdejong** [comment]
Ah, excellent, thanks!
